### PR TITLE
rgw: tooling to locate rgw objects with missing rados components

### PR DESCRIPTION
**Pull request by ivancich**
This has some alterations to RGW and adds two new tools -- rgw-gap-list and rgw-gap-list-comparator. These can be used to help locate, albeit imperfectly, missing rados objects caused by an earlier bug (see #38228).

Signed-off-by: J. Eric Ivancich <ivancich@redhat.com>
Signed-off-by: Michael Kidd <linuxkidd@gmail.com>

Fixes: https://tracker.ceph.com/issues/49063

### CODE REVIEW

**Comment by ivancich** [comment]
@tchaikov The addition of the rgw-gap-list and rgw-gap-list-comparator scripts creates a dependency on gawk, which I feel odd about. Would you please check the build file changes to make sure that's all correct (in addition to any other aspects you'd like to review).

**Comment by ivancich** [comment]
@linuxkidd Do you get Michael? Trying to add you as a reviewer, but so far I cannot.

**Comment by ivancich** [comment]
@linuxkidd, it's been so long since I did any of this. But perhaps you need to request to join the ceph org -- https://github.com/orgs/ceph/people?

**Comment by linuxkidd** [comment]
Looks like I need an org admin to add me... maybe @liewegas knows how or who could do it?

**Comment by ivancich** [comment]
@djgalloway Would you mind adding @linuxkidd to the ceph project, so I can invite him to be a reviewer of this PR? Thanks!!!

**Comment by djgalloway** [comment]
> @djgalloway Would you mind adding @linuxkidd to the ceph project, so I can invite him to be a reviewer of this PR? Thanks!!!

Invited

**Comment by ivancich** [comment]
> > @djgalloway Would you mind adding @linuxkidd to the ceph project, so I can invite him to be a reviewer of this PR? Thanks!!!
> 
> Invited

Thank you!

**Comment by linuxkidd** [comment]
joined...

**Review by linuxkidd — APPROVED** [review]

**Comment by ivancich** [comment]
@ktdreyer , @dmick : I'm looking for a review of the build aspects of this PR -- 2 executable scripts are added (one bash, one awk), and a dependency of rgw on gawk is created. This is not my area of expertise and I've sought help from @tchaikov in the past, but he doesn't seem to be around. Would either/both of you be able to review the PR on that front?

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -3227,6 +3232,8 @@ int main(int argc, const char **argv)
 
   SimpleCmd cmd(all_cmds, cmd_aliases);
 
+  boost::optional<std::string> rgw_obj_fs; // radoslist field separator
```
Comment (tchaikov) [inline src/rgw/rgw_admin.cc:3235]: nit, why not using `std::optional<>`. see https://en.cppreference.com/w/cpp/utility/optional

File: src/rgw/CMakeLists.txt [hunk 2]
```diff
@@ -441,5 +441,5 @@ if(WITH_TESTS)
     DESTINATION bin)
 endif(WITH_TESTS)
 
-install(PROGRAMS rgw-orphan-list
+install(PROGRAMS rgw-orphan-list rgw-gap-list rgw-gap-list-comparator
```
Comment (tchaikov) [inline src/rgw/CMakeLists.txt:444]: nit, could list each of the programs on its own line. like
```cmake
install(PROGRAMS
  rgw-orphan-list
  rgw-gap-list
  rgw-gap-list-comparator
  DESTINATION bin)
```

File: ceph.spec.in [hunk 3]
```diff
@@ -713,6 +713,7 @@ Requires:	librados2 = %{_epoch_prefix}%{version}-%{release}
 Requires:	librgw2 = %{_epoch_prefix}%{version}-%{release}
 %if 0%{?rhel} || 0%{?fedora}
 Requires:	mailcap
+Requires:	gawk
```
Comment (tchaikov) [inline ceph.spec.in:716]: probably we can move it out of the `%if 0%{?rhel} || 0%{?fedora}` block. as I think gawk is very likely available on other distros using rpm packaging system.
also, could use 
```
%if 0%{?weak_deps}
Recommends:	gawk
%endif
```

but I don't feel strong about this.

File: debian/radosgw.install [hunk 4]
```diff
@@ -5,6 +5,8 @@ usr/bin/radosgw-es
 usr/bin/radosgw-object-expirer
 usr/bin/radosgw-token
 usr/bin/rgw-orphan-list
+usr/bin/rgw-gap-list
+usr/bin/rgw-gap-list-comparator
```
Comment (tchaikov) [inline debian/radosgw.install:9]: might want to move these two lines up so the paths are ordered alphabetically.

**Comment by tchaikov** [comment]
@ivancich sorry for the latency.

**Comment by ivancich** [comment]
> @ivancich sorry for the latency.

No problem! I appreciate the insights!

File: ceph.spec.in (hunk 3 above)
Comment (ivancich) [reply ceph.spec.in:716]: I went the weak_deps route.

**Review by tchaikov — APPROVED** [review]
LGTM from the packaging perspective.

**Comment by ivancich** [comment]
I ran a minimal teuthology test over the part that exercises the code that changed:

http://pulpito.ceph.com/ivancich-2021-02-04_19:31:17-rgw-wip-rgw-gap-list-distro-basic-gibba/

**Comment by ivancich** [comment]
@harishmunjulur Sorry that I pulled this out from under you after you started a QA run. I should have marked it with my own testing label.

**Pull request merged**